core/sql/optimizer/NATable.cpp

```diff
@@ -3702,25 +3706,47 @@ NAType* getSQColTypeForHive(const char* hiveType, NAMemory* heap)
       else
        continue;
     }
+
     if(pend - pstart > 0)
     {
+      if( (pend - pstart) >= MAX_NUM_LEN ) // too long
+        return NULL;
       strncpy(pstr,hiveType+pstart, pend-pstart);
       p=atoi(pstr);
     }
+
     if(send - sstart > 0)
     {
+      if( (send - sstart) >= MAX_NUM_LEN ) // too long
+        return NULL;
       strncpy(sstr,hiveType+sstart,send-sstart);
       s=atoi(sstr);
     }
 
-    if( (p>0) && (s>0) )
+    if( (p>0) && (p <= MAX_PRECISION_ALLOWED) ) //have precision between 1 - 18
```

Comment:
See comment above. Also, if we have an invalid scale, we shouldn't take the next if statement. Instead, if the scale is outside the range [0, p], assert or give an error.